File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

int solution(string s) {
    int answer = 0;
    
    for(int i = 0; i<s.length(); i++){
        stack<char> st;
        bool check = true;
        for(int j = 0; j<s.length(); j++){
            if(s[j] == '(' || s[j] == '{' || s[j] == '['){
                st.push(s[j]);
            }
            else if(st.empty()){
                if(s[j] == ')' || s[j] == '}' || s[j] == ']')
                    check = false;
            }
            else{
                if(!st.empty()){
                    if(s[j] == ')' && st.top() == '('){
                        st.pop();
                    }
                    else if(s[j] == '}' && st.top() == '{'){
                        st.pop();
                    }
                    else if(s[j] == ']' && st.top() == '['){
                        st.pop();
                    }
                }
            }
            
        }
        
        if(st.empty() && check) answer++;
        
        char c = s.front();
        s.erase(0,1);
        s.push_back(c);
    }
    
    return answer;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = 0;
        
        if(s.length() % 2 == 1) return 0;
        
        for(int i = 0; i<s.length(); i++) {
            Stack<Character> st = new Stack<>();
            boolean isPossible = true;
            for(int j = 0; j<s.length(); j++){
                int idx = (i + j) % s.length();
                
                if(s.charAt(idx) == '(' || s.charAt(idx) == '[' || s.charAt(idx) == '{')
                    st.push(s.charAt(idx));
                else if(st.isEmpty()) {
                    if(s.charAt(idx) == ')' || s.charAt(idx) == '}' || s.charAt(idx) == ']') {
                        isPossible = false;
                        break;
                    }
                }
                else {
                    if(st.peek() == '(' && s.charAt(idx) == ')')
                        st.pop();
                    else if(st.peek() == '{' && s.charAt(idx) == '}')
                        st.pop();
                    else if(st.peek() == '[' && s.charAt(idx) == ']')
                        st.pop();
                    else {
                        isPossible = false;
                        break;
                    }
                }
            }
            
            if(isPossible) answer++;
        }
        
        return answer;
    }
}